adding dither function

leolca opened the pull request:
I've made an implementation of the dither function, compatible with Matlab implementation (https://www.mathworks.com/help/matlab/ref/dither.html). In Matlab it seems not to be a part of Image package, so I guess it would also fit directly into octave core.

Commit:
adding dither function

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+        print_usage;
+  endif
+  if ndims (RGB) == 2
+    RGB = cat (3, RGB, RGB, RGB); % Duplicate grayscale to RGB
```
For end-of line comments, use two spaces and a single `#` as the comment character.

mmuetzel replied:
Thanks for fixing this in this specific line. But there are many more with the same style issue.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+    endif
+  endif
+  if ndims (RGB) != 3 || size (RGB, 3) != 3
+    error('dither: RGB must be an m x n x 3 array.');
```
Use a whitespace between function name and opening parenthesis.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+  if ndims (RGB) != 3 || size (RGB, 3) != 3
+    error('dither: RGB must be an m x n x 3 array.');
+  end
+  if !ismatrix (map) || size (map, 2) != 3 || min (map(:)) < 0 || max (map(:)) > 1
```
Use a space after the negation operator (i.e., `! ismatrix (map)`).
Use parenthesis around the condition in `if` statements.

mmuetzel replied:
Thanks for fixing this in this specific line. But there are many more with the same style issue.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+  endif
+  Qe = min (Qe, 16); % Cap Qe to avoid excessive precision
+
+  % Scale RGB and map to [0, 1]
```
Use `##` to introduce comments that span the entire line.

mmuetzel replied:
Thanks for fixing this in this specific line. But there are many more with the same style issue.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+## @deftypefn  {Function File} {@var{X} = } dither (@var{RGB}, @var{map})
+## @deftypefnx {Function File} {@var{X} = } dither (@var{RGB}, @var{map}, @var{Qm}, @var{Qe})
+## @deftypefnx {Function File} {@var{BW} = } dither (@var{I})
+##
```
Please add a one-line description of the function if possible.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+##
+## @code{@var{X} = dither (@var{RGB},@var{map})} creates an indexed image 
+## approximation, using the color provided in the colormap, and uses dithering
+## to increase apparent color resolution. Floyd-Steinberg error filter is used: 
```
This sentence is pretty long. Consider splitting it into two sentences.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+## to increase apparent color resolution. Floyd-Steinberg error filter is used: 
+## [   x  7]
+## [3  5  1] / 16
+## It used a raster scan and no weight renormalization at boundaries.
```
Typo? "used" -> "uses"

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+## It used a raster scan and no weight renormalization at boundaries.
+## The default values are used: @var{Qm}=5, and @var{Qe}=8.
+## 
+## Inputs:
```
We usually don't use "section headings" like this. Also note, that texinfo will convert this and the following four lines into one single long line.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,377 @@
+## error calculations in the Floyd-Steinberg error diffusion algorithm. 
+## It controls the precision of the error values that are calculated and
+## propagated during dithering.  If @var{Qe} < @var{Qm}, the error diffusion
+## process may lose precision, therefore dithering cannot be performed, and the
```
Full-stop after "precision".

mmuetzel reviewed (changes requested):
Thank you very much for your contribution.

Sorry it took so long for someone to review your PR. This is just a read-only mirror of the upstream Mercurial repository. (But we try to port PR from here to the upstream repository if possible.)

I haven't checked the new function yet. But I left a few comments about the coding style that is used in Octave.

For the time being, I only took a look at the function documentation and had a very quick look at the function bodies (mostly for style).

I didn't bother to leave a comment at each and every line where the style should be adapted. But I tried to describe the coding style at some examples.

leolca committed:
Update dither.m

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+## error calculations in the Floyd-Steinberg error diffusion algorithm. 
+## It controls the precision of the error values that are calculated and
+## propagated during dithering.  If @var{Qe} < @var{Qm}, the error diffusion
+## process may lose precision. Therefore dithering cannot be performed, and the
```
Two spaces after full stop.

leolca committed:
Refactor comments and formatting in dither.m

Updated comments and formatting in the dither.m file to comply with style guidelines.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -74,10 +74,10 @@
     RGB = cat (3, RGB, RGB, RGB);  # Duplicate grayscale to RGB
     if nargin < 2,
-      map = [0 0 0; 1 1 1]; % binary (black and white) colormap
+      map = [0 0 0; 1 1 1]; # binary (black and white) colormap
```
Use two spaces before inline comments here and anywhere else where applicable.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+    endif
+  endif
+  if (ndims (RGB) != 3 || size (RGB, 3) != 3)
+    error ('dither: RGB must be an m x n x 3 array.');
```
No full stop at the end of error messages here and anywhere else where applicable.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+      for i = 1:size (RGB, 1)
+      for j = 1:size (RGB, 2)
+        X(i, j) = rgb2indLUT (RGB(i, j, :), map, Qm);
+      endfor
```
Indent with two spaces at each level.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+  endif
+  if (nargin < 3)
+    Qm = 5; % Default quantization bits
+  end
```
Use keywords like `endif` instead of `end` here and anywhere else where applicable.

mmuetzel commented:
Thank you again for the modifications.

Please, also add the new file to the build system. (So, it will be included when creating a source tarball). For that, add the new file to the list starting here:
https://github.com/gnu-octave/octave/blob/c77d7580cd7ab01303c06c7b84e9a1935e0c735d/scripts/image/module.mk#L15
(That list is in lexical order afaict.)

Also, please add the documentation for the new file to the manual. A good place might be around here:
https://github.com/gnu-octave/octave/blob/c77d7580cd7ab01303c06c7b84e9a1935e0c735d/doc/interpreter/image.txi#L143-L154
But maybe, you'll find a better location for that.

leolca committed:
Fix error messages, comments, indentation and keywords in dither.m

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+##
+## This program is distributed in the hope that it will be useful,
+## but WITHOUT ANY WARRANTY; without even the implied warranty of
+## MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
```
The "standard" copyright statement in other files has two spaces after the full-stop in this line.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+## @deftypefn  {Function File} {@var{X} = } dither (@var{RGB}, @var{map})
+## @deftypefnx {Function File} {@var{X} = } dither (@var{RGB}, @var{map}, @var{Qm}, @var{Qe})
+## @deftypefnx {Function File} {@var{BW} = } dither (@var{I})
+## Quantize an image, using dithering to increase the apparent color resolution.
```
I'm not a native speaker. But I think, we don't need the comma in this sentence.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+## [   x  7]
+## [3  5  1] / 16
+## It uses a raster scan and no weight renormalization at boundaries.
+## The default values are used: @var{Qm}=5, and @var{Qe}=8.
```
No comma before "and" when listing less than three elements.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+## It uses a raster scan and no weight renormalization at boundaries.
+## The default values are used: @var{Qm}=5, and @var{Qe}=8.
+## 
+## @var{RGB} is a mxnx3 array with values in [0, 1] (double) or [0, 255] (uint8).
```
Did you check how `texinfo` handles this? Do we need something like `@code{[0, 1]}` here?

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+## color axis (R, G, B).  @var{Qm} defines the precision of the color space 
+## discretization used to map input RGB values to those colors available in the
+## colormap.  @var{Qe} is the number of quantization bits for the color space
+## error calculations in the Floyd-Steinberg error diffusion algorithm. 
```
Remove whitespace at end of line here and anywhere else where applicable.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+## Ref [1] Floyd, R. W., and Steinberg, L., An Adaptive Algorithm for Spatial 
+## Gray Scale, International Symposium Digest of Technical Papers, Society for 
+## Information Displays, 1975, p. 36.
+## Ref [2] Ulichney. R., Digital Halftoning, The MIT Press, 1987.
```
Consider using the `@cite` texinfo macro.

leolca committed:
Refactor comments in dither.m

Updated comments for clarity and consistency in the dither function.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+  ##   - Uses a persistent lookup table (LUT) for speed.
+  ##   - LUT is recomputed if map or Qm changes.
+  ##   - Warns if Qm is too large (>8) due to memory constraints.
+  ##   - Assumes input pixel and map are properly scaled (pixel auto-scaled if needed).
```
Even if the help text of sub-functions might not be directly accessible, it might be a good idea to use `texinfo` syntax here for consistency.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+## It uses a raster scan and no weight renormalization at boundaries.
+## The default values are used: @var{Qm}=5 and @var{Qe}=8.
+##
+## @var{RGB} is a mxnx3 array with values in @code{[0, 1]} (double) or @code{[0, 255]} (uint8).
```
Break lines in documentation before 80 (or 74) characters. (Exception are lines with `@deftypefn` or `@deftypefnx`.)

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,379 @@
+    error ('rgb2indLUT: Colormap must be a c-by-3 matrix');
+  endif
+  if (nargin < 3)
+    Qm = 5; % Default quantization bits
```
I thought that you fixed that at some point:
Inline comments should use `#`.
There should be two spaces before `#` for inline comments.

There are a couple more inline comments with `%` a few lines down.

leolca committed:
Refactor comments and formatting in dither.m

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,413 @@
+  if (! ismatrix (map) ||  ...
+      size (map, 2) != 3 || ...
+      min (map(:)) < 0 || ...
+      max (map(:)) > 1)
```
No need to break this condition as often. And if you break conditions like that, please break them *before* (Boolean) operators.
E.g.:
```
  if (! ismatrix (map) || size (map, 2) != 3 ...
      || min (map(:)) < 0 || max (map(:)) > 1)
```

mmuetzel commented:
Thank you again for all these nitty changes.

Did you read the above [comment](https://github.com/gnu-octave/octave/pull/33#issuecomment-3421384056)?
Are you comfortable making those changes?

leolca committed:
Refactor colormap validation condition

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+  endif
+  if (nargin < 3)
+    Qm = 5;  # Default quantization bits
+  endif
```
This check shouldn't be necessary because you already set a default value in the input arguments list.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+endfunction
+
+function [X, Y, Z] = ndims_grid (nx, ny, nz)
+  ## NDIMS_GRID Create 3D grid indices (emulates meshgrid for 3D).
```
``` suggestion
  ## Create 3D grid indices (emulates meshgrid for 3D).
```

mmuetzel replied:
Same as above. Delete the previous line

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+## Information Displays, 1975, p. 36.}
+## @cite{Ulichney, R., Digital Halftoning, The MIT Press, 1987.}
+##
+## @seealso{rgb2ind, imapprox}
```
Afaics, `imapprox` is (currently) implemented as part of the "image" package. It doesn't exist in core Octave. So, this will likely lead to an error when trying to build the documentation.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+## -*- texinfo -*-
+## @deftypefn  {Function File} {@var{X} = } dither (@var{RGB}, @var{map})
+## @deftypefnx {Function File} {@var{X} = } dither (@var{RGB}, @var{map}, @var{Qm}, @var{Qe})
+## @deftypefnx {Function File} {@var{BW} = } dither (@var{I})
```
``` suggestion
## @deftypefn  {} {@var{X} = } dither (@var{RGB}, @var{map})
## @deftypefnx {} {@var{X} = } dither (@var{RGB}, @var{map}, @var{Qm}, @var{Qe})
## @deftypefnx {} {@var{BW} = } dither (@var{I})
```

mmuetzel replied:
It looks like you forgot to delete the previous lines.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+## dithering to increase apparent color resolution.  Floyd-Steinberg error
+## filter is:
+## @code{[   x  7]}
+## @code{[3  5  1] / 16}
```
This part renders as the following as PDF:

<img width="1107" height="73" alt="Image" src="https://github.com/user-attachments/assets/6a15638a-abb6-4c4a-9505-d01ff44c2eaa" />

Is that what you intended here?

mmuetzel replied:
With the current iteration, this looks like this in the PDF:
<img width="1127" height="273" alt="image" src="https://github.com/user-attachments/assets/50489178-dd2b-4b66-a412-923c088e3230" />

Like this in the console output:
```
     ‘X = dither (RGB,MAP)’ creates an indexed image approximation.  It
     uses the color provided in the colormap, and uses dithering to
     increase apparent color resolution.  Floyd-Steinberg error filter
     is:
               x    7/16
          3/16 5/16 1/16

     It uses a raster scan and no weight renormalization at boundaries.
     The default values are used: QM=5 and QE=8.
```
And like this in the Qt documentation browser:
<img width="882" height="127" alt="image" src="https://github.com/user-attachments/assets/72948839-69b7-4fe0-8471-736fe06ae6e6" />

That looks reasonable to me now. 👍

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+## It uses a raster scan and no weight renormalization at boundaries.
+## The default values are used: @var{Qm}=5 and @var{Qe}=8.
+##
+## @var{RGB} is a mxnx3 array with values in @code{[0, 1]} (double) or
```
In other files (e.g., `ind2rgb`), capital letters are used for dimensions. E.g. `@nospell{MxNx3}`.

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+## @end example
+##
+## @code{@var{X} = dither (@var{RGB}, @var{map}, @var{Qm}, @var{Qe})}
+##
```
This block looks a bit out of place.
Should it be removed?

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+## @cite{Floyd, R. W., and Steinberg, L., An Adaptive Algorithm for Spatial Gray
+## Scale, International Symposium Digest of Technical Papers, Society for
+## Information Displays, 1975, p. 36.}
+## @cite{Ulichney, R., Digital Halftoning, The MIT Press, 1987.}
```
This renders like the following as PDF:

<img width="1147" height="115" alt="Image" src="https://github.com/user-attachments/assets/27b7c13e-e70e-4b97-85be-98ea97799666" />

Maybe something along the lines of what is done in `condest.m` would look better?
https://github.com/gnu-octave/octave/blob/81aa7937c6a42a079702654b4f0daef75928665f/scripts/linear-algebra/condest.m#L98-L111

mmuetzel replied:
This looks like this in the PDF now:
<img width="1107" height="201" alt="image" src="https://github.com/user-attachments/assets/8ed05668-2d3b-4825-ad98-b85f641c7acb" />

Like this in the console:
```
     References:

        • Floyd, R. W. and Steinberg, L., ‘An Adaptive Algorithm for
          Spatial Gray Scale’.  International Symposium Digest of
          Technical Papers, Society for Information Displays, 1975, p.
          36.

        • Ulichney, R., ‘Digital Halftoning’.  The MIT Press, 1987.
```

And like this in the Qt documentation browser.
<img width="875" height="96" alt="image" src="https://github.com/user-attachments/assets/8ed73351-0f5a-4b31-a787-57085fb0d87d" />

Looks good to me. 👍

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,411 @@
+%! # Lenna
+%! url = ['https://upload.wikimedia.org/wikipedia/en/7/7d/' ...
+%!       'Lenna_%28test_image%29.png'];
+%! rgb_image = imread(url);
```
Use space between function name and opening parenthesis.

mmuetzel commented:
I finally came around to build Octave and its documentation with your changes locally.

All demos and tests pass for me. 👍

I made the following additional changes for integrating the new file into Octave:
``` diff
diff --git a/doc/interpreter/image.txi b/doc/interpreter/image.txi
--- a/doc/interpreter/image.txi
+++ b/doc/interpreter/image.txi
@@ -153,6 +153,10 @@ formats.
 
 @DOCSTRING(ind2rgb)
 
+The following function allows processing an existing image.
+
+@DOCSTRING(dither)
+
 Octave also provides tools to produce and work with movie frame structures.
 Those structures encapsulate the image data (@qcode{"cdata"} field) together
 with the corresponding colormap (@qcode{"colormap"} field).

diff --git a/scripts/help/__unimplemented__.m b/scripts/help/__unimplemented__.m
--- a/scripts/help/__unimplemented__.m
+++ b/scripts/help/__unimplemented__.m
@@ -747,7 +747,6 @@ function rlist = missing_functions ()
   "discretize",
   "dissect",
   "distances",
-  "dither",
   "docsearch",
   "dragrect",
   "duration",

diff --git a/scripts/image/module.mk b/scripts/image/module.mk
--- a/scripts/image/module.mk
+++ b/scripts/image/module.mk
@@ -25,6 +25,7 @@ FCN_FILE_DIRS += \
   %reldir%/cool.m \
   %reldir%/copper.m \
   %reldir%/cubehelix.m \
+  %reldir%/dither.m \
   %reldir%/flag.m \
   %reldir%/frame2im.m \
   %reldir%/getframe.m \
```

I also had to make the following change or the documentation wouldn't build:
``` diff
diff --git a/scripts/image/dither.m b/scripts/image/dither.m
--- a/scripts/image/dither.m
+++ b/scripts/image/dither.m
@@ -71,7 +71,7 @@
 ## Information Displays, 1975, p. 36.}
 ## @cite{Ulichney, R., Digital Halftoning, The MIT Press, 1987.}
 ##
-## @seealso{rgb2ind, imapprox}
+## @seealso{rgb2ind}
 ## @end deftypefn
 
 function X = dither (RGB, map, Qm = 5, Qe = 8)
```

Could you please integrate those changes (and address the other comments that I left in the last round of review)?

leolca committed:
Fix documentation for dither function

Updated documentation for the dither function to correct matrix dimensions and improve clarity.

leolca commented:
thanks for the feedback, suggestions and comments

leolca committed:
Remove duplicate dither function documentation

mmuetzel commented:
Thanks again for the follow-up. Could you please integrate the changes to the three files mentioned in the [comment above](https://github.com/gnu-octave/octave/pull/33#issuecomment-3433175247)?
Without them, I need to make those changes manually each time I try to build locally with the changes from this PR.

leolca committed:
Merge branch 'gnu-octave:default' into dither

Commit:
doc files update

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,424 @@
+## Example:
+## @example
+## X = dither (RGB, map);
+## @end example
```
The example is a bit out of place in the middle of the documentation.
Would it make sense to move it to the end of the documentation? (Just before the references.)

Also, it is not much more than a repetition of one of the function prototypes as they are already mentioned in the `deftypefn` lines at the very start of the documentation. So, it doesn't add much value.
Should it better be removed entirely?

mmuetzel commented on scripts/image/dither.m:
```diff
@@ -0,0 +1,424 @@
+## is a black and white image where dithering creates the illusion of shades of
+## gray.
+##
+## @itemize
+## @item
```
Would it make sense to introduce this with a "heading" like in the other file?
``` suggestion
 ## References: 
 ## 
 ## @itemize 
 ## @item 
```

Commit:
remove demo and add reference heading

mmuetzel commented:
I squashed the changes from this PR into a single commit, added a commit message, and pushed it with very minor changes to the upstream Mercurial repository.
It should be appearing in this mirror shortly.

Thank you again very much for your contribution.

mmuetzel commented:
Synced here:
https://github.com/gnu-octave/octave/commit/c014cb1f1dd3a5f3c0621f8af007326275ef6a3f
